docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+# Record and Replay ADR
+
+## Submitters
+- Lenny Goodell (Intel)
+
+## Change Log
+- [Pending](https://github.com/edgexfoundry/edgex-docs/pull/863) (2022-09-20)
+
+## Referenced Use Case(s)
+- [Record and Replay UCR](https://docs.edgexfoundry.org/2.3/design/ucr/Record-and-Replay/)
+
+## Context
+This design involves creating a new Application Service that is responsible for the requirements in the above referenced UCR. This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.
+
+## Proposed Design
+A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. 
```

Comment:
Why should this be a service and have a RESTful API to control the operations? Why not make this an application which takes CLI sub-commands (record, replay, import, export) and arguments or a configuration file and can be gracefully shutdown on request? **Why can't this be simply added to the EdgeX CLI?** The UCR did not mention the need for a service or list any requirement for network-based / remote control of the operations.

This could be justified if there was a need for parallel recording sessions.

A application is superior because it can also offer logging to the user to inform them about the progress.

Reply:
See my response to your same general comment.

Reply:
Your [comment](https://github.com/edgexfoundry/edgex-docs/pull/863#issuecomment-1261328653) does not answer any of my questions.

Reply:
Not sure what you are looking for here.  Correct the UCR doesn't have a requirement of service via cli, that is a design choice.

> This could be justified if there was a need for parallel recording sessions.

This could be a future enhancement

> A application is superior because it can also offer logging to the user to inform them about the progress.

Yes, this is correct. Expect implementation to have logging to show progress.

Reply:
CLI vs networking API isn't a design choice. CLI is a user-facing interface of an application, while a networking API for machines to communicate with a web service. They serve very different purposes and have different use cases.

If we are offering a web service, the progress info is only useful if it is available also over the networking API; for example, an API call to query the progress.

Reply:
> the progress info is only useful if it is available also over the networking API

Yes, I agree that status endpoint would be very useful. I have added them.

Reply:
> CLI vs networking API isn't a design choice.

How can it not be a design choice. IMO it is a major design choice. UCR doesn't dictate it be designed using CLI or REST API.

Reply:
The design could call for a RESTful service or a CLI, that address the requirements, thus it is a design decision.

Reply:
> Why should this be a service

This has to be a **service** as it must connect  to the MessageBus and consume Events over a long period of time (just like other App Services). A CLI would have to duplicate this long run service in the background which is not very CLI'ish.

Reply:
> > Why should this be a service
> 
> This has to be a **service** as it must connect to the MessageBus and consume Events over a long period of time (just like other App Services). A CLI would have to duplicate this long run service in the background which is not very CLI'ish.

Okay, at last! That is a compelling reason which should be documented in the design.

Reply:
Added to `Proposed Design` section and updated CLI item in `Considerations` section.